data/tutorials/language/0lg_09_env.md

```diff
@@ -0,0 +1,675 @@
+In order to perform iterated computations, a function may call itself. Such a function is called _recursive_. In OCaml, recursive functions must be defined and explicitly declared by using `let rec`. It is not possible to accidentally create recursion loops between functions. As a consequence, recursive functions can't be anonymous.
+
+The classic (and very inefficient) way to present recursion is using the function that computes the [Fibonacci Sequence](https://en.wikipedia.org/wiki/Fibonacci_sequence).
+```ocaml
+# let rec fibo n = if n <= 1 then n else fibo (n - 1) + fibo (n - 2);;
+val fibo : int -> int = <fun>
+
+# let u = List.init 10 Fun.id;;
+val u : int list = [0; 1; 2; 3; 4; 5; 6; 7; 8; 9]
+
+# List.map fibo u;;
+- : int list = [0; 1; 1; 2; 3; 5; 8; 13; 21; 34]
+```
+
+This version of `fibo` is inefficient because the number of recursive calls created doubles at each call, which creates exponential growth.
+
+> Note: `List.init` is a standard library function that allows you to create a list by applying a given function to a sequence of integers, and `Fun.id` is the identity function, which returns its argument unchanged. We created a list with the numbers 0 - 9 and named it `xs`. We applied the `fibo` function to every element of the list using `List.map`.
+
+This version does a better job:
+```ocaml
+# let rec fib m n i = if i = 0 then m else fib n (n + m) (i - 1);;
+val fib : int -> int -> int -> int = <fun>
+
+# let fib = fib 0 1;;
+val fib : int -> int = <fun>
+
+# List.init 10 Fun.id |> List.map fib;;
+- : int list = [0; 1; 1; 2; 3; 5; 8; 13; 21; 34]
+```
+
+The first version (`fibo`) takes two extra parameters: the two previously computed Fibonacci numbers.
+
+The second version (`fib`) uses the first two Fibonacci numbers as initial values. There is nothing to be computed when returning from a recursive call, so this enables the compiler to perform an optimisation called “tail call elimination.” This turns recursivity into imperative iteration in the generated native code and leads to improved performances.
+
+> Note: Notice that the `fib` function has three parameters (`m n i`) but only two arguments were passed (`0 1`). This is called *partial application*.
+
+## Multiple Arguments Functions
+
+### With Syntactic Sugar
+
+To define a function with multiple arguments, each must be listed between the name of the function (right after the `let` keyword) and the equal sign, separated by space. Here is an example:
+```ocaml
+# let sweet_cat x y = x ^ " " ^ y;;
+val sweet_cat : string -> string -> string = <fun>
+
+# sweet_cat "kitty" "cat";;
+- : string = "kitty cat"
+```
+
+This is how most multiple-argument functions are defined. Alternatives exist, but this should be the default.
+
+### Without Syntactic Sugar
+
+The function `sweet_cat` is the same as this one:
+```ocaml
+# let sour_cat = fun x -> fun y -> x ^ " " ^ y;;
+val sour_cat : string -> string -> string = <fun>
+
+# sweet_cat "kitty" "cat";;
+- : string = "kitty cat"
```

Comment:
```suggestion
# sour_cat "kitty" "cat";;
- : string = "kitty cat"
```

Should this be `sour_cat`, as that's what we just defined? Or did you mean to apply the previous function?